### ISSUE TITLE
Serializing and deserializing Task<T> to json into Redis cache awaits a never returning task 

### ISSUE THREAD

**Issue by erikwuu** [nopSolutions/nopCommerce#3403]
nopCommerce version: 4.0 

When enabling the Azure blob storage and Redis cache, the Redis cache manager awaits a deserialized task which will never return. After a bunch of debugging sessions, I've found that the method GeneratedThumbExistsAsync in the AzurePictureService never returns. This will eventually cause a timeout error on the home page and the product admin panel since those pages eventually call this method.

On startup, the Redis cache manager Get method inserts the thumbnail keys into Redis, saving the task object as value. When trying to deserialize the task, it awaits the deserialization. It happens in the following piece of code: (which can be found in the method GeneratedThumbExistsAsync in the AzurePictureService)

 ```
return await  _cacheManager.Get(key, async () =>
                {
                    //GetBlockBlobReference doesn't need to be async since it doesn't contact the server yet
                    var blockBlob = _container.GetBlockBlobReference(thumbFileName);

                    return await blockBlob.ExistsAsync();
                });
```
The fix that I've applied is to execute the acquire function synchronously, however, this will have a performance impact since the call is being made synchronously to Azure blob storage.

```
 return  _cacheManager.Get(key, () =>
                {
                    //GetBlockBlobReference doesn't need to be async since it doesn't contact the server yet
                    var blockBlob = _container.GetBlockBlobReference(thumbFileName);

                    return blockBlob.ExistsAsync().ConfigureAwait(false).GetAwaiter().GetResult();
                });
```

I've also looked into the source code of nop 4.1, which also should have this problem. I'm curious whether I'm the only one having this problem since this should work out of the box.

Steps to reproduce the problem: Enable Azure blob storage with Redis. Go to the home page or product admin panel, these pages will eventually time out.

**Comment by erikwuu** [comment]
The fix I've applied to still get everything async (for nop 4.00)

Nop.Core -> Caching -> Extensions.cs

```
public static async Task<T> GetAsync<T>(this IStaticCacheManager cacheManager, string key, Func<Task<T>> acquire)
        {
            return await GetAsync(cacheManager, key, DefaultCacheTimeMinutes, acquire);
        }


        public static async Task<T> GetAsync<T>(this IStaticCacheManager cacheManager, string key, int cacheTime, Func<Task<T>> acquire)
        {
            //item already is in cache, so return it
            if (cacheManager.IsSet(key))
                return cacheManager.Get<T>(key);

            //or create it using passed function
            var result = await acquire();

            //and set in cache (if cache time is defined)
            if (cacheTime > 0)
                cacheManager.Set(key, result, cacheTime);

            return result;
        }
```

And in Nop.Services -> Media -> AzurePictureService.cs

```
        protected virtual async Task<bool> GeneratedThumbExistsAsync(string thumbFilePath, string thumbFileName)
        {
            try
            {
                var key = string.Format(THUMB_EXISTS_KEY, thumbFileName);
                return await _cacheManager.GetAsync(key, async () =>
                {
                    //GetBlockBlobReference doesn't need to be async since it doesn't contact the server yet
                    var blockBlob = _container.GetBlockBlobReference(thumbFileName);

                    return await blockBlob.ExistsAsync();
                });
            }
            catch { return false; }
        }
```

However, the cachemanager Get method should be replaced in all locations where the Func<T> acquire parameter is async.

**Comment by skoshelev** [comment]
@erikwuu Thank you for your help and investigate for this issue, we made changes without extension method but using your code.

Closed #3403.

**Comment by Leftyx** [comment]
@skoshelev, I made some changes to my repository some time ago and submitted a [PR](https://github.com/nopSolutions/nopCommerce/pull/3404) with the changes.
I had worked with similar problems in the past and you will find some changes which you might want to consider.
It's working fine in production.

**Comment by skoshelev** [comment]
Closed #3403
